File: cs/3_2_1_state.cs
using System;
using System.Collections.Generic;

namespace Behavioral.State
{

    interface IPass
    {
        string Pass(int rating = 0);
    }
    class AutomatePass : IPass
    {
        public string Pass(int rating)
        {
            if (rating >= 90)
                return "A";
            if (rating >= 82)
                return "B";
            if (rating >= 74)
                return "C";
            if (rating >= 64)
                return "D";
            return "E";
        }
    }

    class NormalPass : IPass
    {
        public string Pass(int rating = 0)
        {
            Console.WriteLine("Take a examination ticket... ");
            //...
            return "E";
        }
    }

    class ExclusionPass : IPass
    {
        public string Pass(int rating = 0)
        {
            Console.WriteLine("Non-admission to the exam");
            return "F";
        }
    }

    class SubjectMark
    {
        public string Name { get; set; }
        private IPass _passState;
        private int _rating;
        public int Rating
        {
            get
            {
                return _rating;
            }
            set
            {
                _passState = GetPassState(value);
                _rating = value;
            }
        }
        public static List<Tuple<Predicate<int>, IPass>> state_selectors = new List<Tuple<Predicate<int>, IPass>>() {
            new (
                rating => 60<=rating && rating<=100,
                new AutomatePass()
                ),
            new (
                rating => 35<=rating && rating<60,
                new NormalPass()
                ),
            new (
                rating => 0<=rating && rating<35,
                new ExclusionPass()
                ),
        };

        private IPass GetPassState(int value)
        {
            var state_selector = state_selectors.Find(s => s.Item1(value));
            if (state_selector == null)
                throw new Exception($"{value} is not valid");
            return state_selector.Item2;
        }

        public SubjectMark(string Name, int Rating)
        {
            this.Name = Name;
            this.Rating = Rating;
        }

        public void Pass()
        {
            Console.WriteLine("===========================");
            Console.WriteLine($"Passing exam {Name}");
            string mark = _passState.Pass(Rating);
            Console.WriteLine($"Your mark is {mark}");
        }

    }
}

File: python/patterns/_3_2_1_state.py
from typing import Protocol, Callable


class IPass(Protocol):
    def try_pass(self, rating: int) -> str: ...


class AutomatePass:  # (IPass)
    def try_pass(self, rating: int) -> str:
        if (rating >= 90):
            return "A"
        elif (rating >= 82):
            return "B"
        elif (rating >= 74):
            return "C"
        elif (rating >= 64):
            return "D"
        return "E"


class NormalPass:  # (IPass)
    def try_pass(self, rating: int) -> str:
        print("Take a examination ticket... ")

        return "E"


class ExclusionPass:  # (IPass)
    def try_pass(self, rating: int) -> str:
        print("Non-admission to the exam")

        return "F"


class SubjectMark:
    def __init__(self, name: str, rating: int) -> None:
        self.name: str = name
        self._rating: int = rating
        self._state: IPass = self.get_state(rating)

    @property
    def rating(self) -> int:
        return self._rating

    @rating.setter
    def rating(self, value: int) -> None:
        self._state = self.get_state(value)
        self._rating = value

    state_selectors: list[
        tuple[
            Callable[[int], bool],
            IPass
        ]
    ] = [
        (
            lambda rating: 60 <= rating <= 100,
            AutomatePass()
        ),
        (
            lambda rating: 35 <= rating < 60,
            NormalPass()
        ),
        (
            lambda rating: 0 <= rating < 35,
            ExclusionPass()
        )
    ]

    def get_state(self, rating: int) -> IPass:
        for predicate, pass_state in self.state_selectors:
            if predicate(rating):
                return pass_state
        raise ValueError
    
    def pass_exam(self)->None:
        print(f"Passing exam {self.name}")
        mark:str =   self._state.try_pass (self.rating)
        print(f"Your mark is {mark}")
